ecs-agent/tmds/handlers/v4/handlers.go

```diff
@@ -0,0 +1,99 @@
+// permissions and limitations under the License.
+package v4
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger/field"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/metrics"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/tmds/handlers/utils"
+	state "github.com/aws/amazon-ecs-agent/ecs-agent/tmds/handlers/v4/state"
+
+	"github.com/gorilla/mux"
+)
+
+// v3EndpointIDMuxName is the key that's used in gorilla/mux to get the v3 endpoint ID.
+const (
+	EndpointContainerIDMuxName = "endpointContainerIDMuxName"
+	version                    = "v4"
+)
+
+// ContainerMetadataPath specifies the relative URI path for serving container metadata.
+func ContainerMetadataPath() string {
+	return "/v4/" + utils.ConstructMuxVar(EndpointContainerIDMuxName, utils.AnythingButSlashRegEx)
+}
+
+// ContainerMetadataHandler returns the HTTP handler function for handling container metadata requests.
+func ContainerMetadataHandler(
+	agentState state.AgentState,
+	metricsFactory metrics.EntryFactory,
+) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		endpointContainerID := mux.Vars(r)[EndpointContainerIDMuxName]
+		containerMetadata, err := agentState.GetContainerMetadata(endpointContainerID)
+		if err != nil {
+			logger.Error("Failed to get v4 container metadata", logger.Fields{
+				field.TMDSEndpointContainerID: endpointContainerID,
+				field.Error:                   err,
+			})
+
+			responseCode, responseBody := getContainerErrorResponse(endpointContainerID, err)
+			utils.WriteJSONResponse(w, responseCode, responseBody, utils.RequestTypeContainerMetadata)
+
+			if utils.Is5XXStatus(responseCode) {
+				metricsFactory.New(metrics.InternalServerErrorMetricName).Done(err)()
+			}
+
+			return
+		}
+
+		logger.Info("Writing response for v4 container metadata", logger.Fields{
+			field.TMDSEndpointContainerID: endpointContainerID,
+			field.Container:               containerMetadata.ID,
+		})
+		utils.WriteJSONResponse(w, http.StatusOK, containerMetadata, utils.RequestTypeContainerMetadata)
+	}
+}
+
+// Returns an appropriate HTTP response status code and body for the error.
+func getContainerErrorResponse(endpointContainerID string, err error) (int, string) {
+	var errLookupFailure *state.ErrorLookupFailure
+	if errors.As(err, &errLookupFailure) {
+		return http.StatusNotFound, fmt.Sprintf("V4 container metadata handler: %s",
+			errLookupFailure.ExternalReason())
+	}
+
+	var errMetadataFetchFailure *state.ErrorMetadataFetchFailure
+	if errors.As(err, &errMetadataFetchFailure) {
+		return http.StatusInternalServerError, errMetadataFetchFailure.ExternalReason()
+	}
+
+	logger.Error("Unknown error encountered when handling container metadata fetch failure",
+		logger.Fields{field.Error: err})
+	return http.StatusInternalServerError, "failed to get container metadata"
```

Comment:
nit: Is there any reason we can't return `errMetadataFetchFailure.Error()` instead of hard-coded "failed to get container metadata"?

And separately is there any reason why we don't want to include the contents/info of external reason in the return value of this case? (I might be missing something obvious here)

Reply:
This line is reached when `err` did not contain a `ErrorMetadataFetchFailure`, so `errMetadataFetchFailure` is empty when this line is executed. 

This is the case when the handler does not recognize the error returned by the Agent so it logs the error returned and returns a generic "failed to get container metadata" response to the user.